README.md
# LDmicro 
LDmicro is a program for creating, developing and editing [ladder diagrams](https://en.wikipedia.org/wiki/Ladder_logic),
simulation of a ladder diagram work
and the compilation of ladder diagrams into the native hexadecimal firmware code of the Atmel AVR and Microchip PIC controllers.

Official [LDmicro Ladder Logic for PIC and AVR Home Page](http://cq.cx/ladder.pl)  
Official [LDmicro Forum](http://cq.cx/ladder-forum.pl)  
Actual [manual.txt](https://github.com/LDmicro/LDmicro/blob/master/ldmicro/manual.txt)  

Download executable binaries buildXXXX.zip from the [Latest release](https://github.com/LDmicro/LDmicro/releases/latest)  
You probably need to install the Microsoft Visual C ++ Redistributable Package if Visual C ++ is not installed in your operating system. See [MSVCP100.dll is missing error](https://github.com/LDmicro/LDmicro/issues/174)  

Read Ldmicro [Wiki](https://github.com/LDmicro/LDmicro/wiki) and [HOW TO:](https://github.com/LDmicro/LDmicro/wiki/HOW-TO)  

### Building LDmicro

#### Building with make

LDmicro is built using the Microsoft Visual C++ compiler. If that is
installed correctly, then you should be able to just run

    make.bat

and see everything build.

Various source and header files are generated automatically. The perl
scripts to do this are included with this distribution, but it's necessary
to have a perl.exe in your path somewhere.

The makefile accepts an argument, D=LANG_XX, where XX is the language
code. make.bat supplies that argument automatically, as LANG_EN (English).

#### Building with Cmake

For building LDmicro with Cmake you need Cmake itself, Perl interpreter and C++11 compiler.

You should use out-of-source-tree builds, so create e.g. a directory
`build` in the main ldmicro directory. (If you choose to build in other
directory, replace the `..` in the following instructions with path pointing
to the root/ldmicro directory.)

To build with MSYS + [mingw-w64](http://mingw-w64.org) + Make:
```sh
$ mkdir build
$ cd build
$ cmake -G "MSYS Makefiles" ..
$ make
```

To build with MSYS + mingw-w64 + [Ninja](http://martine.github.io/ninja):
```sh
$ mkdir build
$ cd build
$ cmake -G "Ninja" ..
$ ninja
```

To build within [MSYS2](http://msys2.github.io), make sure you have these
MSYS2 packages installed:
* `make`
* `mingw-w64-i686-gcc`, `mingw-w64-i686-cmake`

#### Build with Microsoft Visual Studio 2017

Visual Studio 2017 supports CMake build system, so you may just follow these
instructions.
1. Start Visual Studio 2017.
2. In menu File, choose submenu Open and Folder.
3. In the open dialog, navigate to LDmicro/ldmicro folder and open it.
4. In menu CMake, choose to Build all.

#### Build with Older Version of Microsoft Visual Studio

To build with older Microsoft Visual Studio 2013 or 2015, you have to
generate project files manually:
```sh
$ mkdir build
$ cd build
$ cmake -G "Visual Studio 12 2013" ..           # MSVC 2013, 32-bit build
$ cmake -G "Visual Studio 12 2013 Win64" ..     # MSVC 2013, 64-bit build
$ cmake -G "Visual Studio 14 2015" ..           # MSVC 2015, 32-bit build
$ cmake -G "Visual Studio 14 2015 Win64" ..     # MSVC 2015, 64-bit build
$ cmake -G "Visual Studio 15 2017" ..           # MSVC 2017, 32-bit build
$ cmake -G "Visual Studio 15 2017 Win64" ..     # MSVC 2017, 64-bit build
$ cmake -G "Visual Studio 16 2019" ..           # MSVC 2019, 32-bit build
$ cmake -G "Visual Studio 16 2019 Win64" ..     # MSVC 2019, 64-bit build
```
Then open the generated solution file `build/ldmicro.sln` in Visual Studio and
build the target `ALL_BUILD`.

You can also choose LDmicro language. For that, you should set LDLANG variable for Cmake.
LDLANG can be one of those values: EN, DE, ES, FR, IT, PT, TR, RU, JA or ALL.
```sh
cmake -G "Ninja" -DLDLANG=DE ..
```


ldmicro/CHANGES.txt
== Release 5.4.2.0

    * Fixed error in V5411, after simulation(edit).

    * Fixed PORTF output pins on "Atmel AVR ATmega2560 100-TQFP"

    * New:  Instruction FRMT STR TO CHAR - string to char iterator.
    It puts one char per cycle into the general variable "char".
    Then "char" variable may send to UART or I2C or SPI etc.
    (Sponsored by OnosTech, thank you very much.)

    * New:  Instruction VAR TO CHAR - varable to char/byte iterator.
    It puts one char/byte per cycle into the general variable "char".
    Then "char" variable may send to UART or I2C or SPI etc.

    * See YouTube OnosTech channel: LDmicro Tutorial Series
    https://www.youtube.com/channel/UCPzl5JwVpGQCCup5YDP0mPw


== Release 5.4.1.1

    * Fixed: Reset the Pull-up resistors when you start a new program. Fix #222

    * Added PWM on Atmega128. Fix #221

    * Fixed: BUS TRASER for PICs MCU.

    * Wiki: BUS TRASER
    https://github.com/LDmicro/LDmicro/wiki/BUS-TRASER


== Release 5.4.1.0

    * Added "Microchip PIC12F1840 8-pin packages" (Sponsored by MGP, thank you very much.)

    * New:  Instruction  ODR: ONE DROP if RISING is the inversion
      of the instruction OSR: ONE-SHOT RISING,
                                _   _   _
                           --[_/ ODR \_/ ]--     ONE DROP if RISING
                                _     _
                           --[_/ OSR_/ \_]--     ONE-SHOT RISING

    * Rename: Instruction OSL: ONE-SHOT LOW LEVEL
      to new name     ODF: ONE DROP if FALLING
                              _     _   _
                           --[ \_ODF \_/ ]--     ONE DROP if FALLING
                              _       _
                           --[ \_OSF_/ \_]--     ONE-SHOT FALLING

    * Wiki: OSR, OSF, ODR, ODF, OSC
    https://github.com/LDmicro/LDmicro/wiki/OSR,-OSF,-ODR,-ODF,-OSC

    * Fixed error: 'true' undeclared (first use in this function). Issue BuildAVR.bat fixed #227
        while(true)   replaced at   while(1)

    * Fixed: BUS TRACER instruction.

    * Wiki: HOW TO: 7 segment display (7SEG & BUS)
    https://github.com/LDmicro/LDmicro/wiki/HOW-TO:-7-segment-display,-7SEG-&-BUS

    * Fix: WIP: SPI Atmel AVR-GCC for ATmega2560.


== Release 5.4.0.5

    * Refactor EEPROM RERSIST_ELEM for AVR

    * Fixed bug: LDmicro Forum - A problem with substracting 2 values #214
      ELEM_ADD & ELEM_SUB refactoring

    * Added "Atmel AVR ATmega1284 44-Pin packages"

    * Fixed PWM for AVR ATmega1284. Six PWM Channels are available.

    * Fixed AVR ATtiny85


== Release 5.4.0.4

    * Fixed the bug in ladder.h generated by version 5.4.0.x when compiling for AVR-GCC (double include).

    * PULSER instruction tested with ATmega2560 (both HEX->ASM and AVR-GCC compilers)
      PULSER instruction tested with PIC16F887 (both HEX->ASM and HI-TECH C compilers)
      See: https://github.com/LDmicro/LDmicro/wiki/PULSER

    * Fixed: Convert manual-es.txt from 65001 UTF-8 code page to 1252 code page


== Release 5.4.0.3

    * Modified menu Settings

    * Fixed lang-fr.txt

    * Some minor hotfixes


== Release 5.4.0.2

    * Fixed Microchip PIC16F887 44-TQFP UART, ISP,I2C pins.

    * Fixed Quad Encoder Output dialog Dir pin name.

    * Fixed Atmel AVR ATmega32U4 pins mapping for Arduino.

    * Fixed Spanish locale. Thank you very much, Paulino Cano.
      lang-es.txt, manual-es.txt updated.


== Release 5.4.0.1 String type support

    * Added
        "Atmel AVR ATmega32U4 44-Pin packages",
        "Atmel AVR ATmega32 44-Pin packages",
        "Atmel AVR ATmega328 32-Pin packages",
        "Atmel AVR ATtiny85 8-Pin packages",
        "Atmel AVR ATtiny85 20-Pin packages",
        "Microchip PIC16F1512 28-Pin SPDIP, SOIC, SSOP",
        "Microchip PIC16F1516 28-Pin SPDIP, SOIC, SSOP",
        "Microchip PIC16F1527 64-Pin packages",
        "Microchip PIC16F1824 14-Pin PDIP, SOIC, TSSOP",
        "Microchip PIC16F1827 18-Pin PDIP, SOIC",
        "Microchip PIC10F202 6-SOT",
        "Microchip PIC16F887 44-TQFP"

    * Added LIBRARIES_FOR\Pic16\UartLib.*
            LIBRARIES_FOR\AVR\UartLib.*

    * Added "9-SEGMENT FONT converter"
           "14-SEGMENT FONT converter"
           "16-SEGMENT FONT converter"
        * Added "BUS tracer" support

    * Added STRING type support in generated C code
        Tested ATmega2560,

    * Added Blue color scheme by Jose GILLES

    * Added by Jose GILLES
        "Microchip PIC18F4520 40-PDIP"

== Release 5.3.1.0

    * Fixed flashMcu.bat and LIBRARIES_FOR ARM,AVR, PIC by Jose GILLES

    * Modified "St ARM STM32F40X 144-LQFP" by Jose GILLES

    * Added "St ARM STM32F10X 48-LQFP / Bluepill" by Jose GILLES

    * Fixed OSR, OSF, OSL, OSC show error.

    * Fixed AVR compilation stack overflow error.

    * Fixed flashMcu.bat for AVR

    * Fixed LUT index range in simulation.

    * Fixed PIC UART baud rate calculation.

    * Fixed PICs PORTA outputs still stay in analog mode in generated C source code.
      Added setPortDigitalIO()

    * Fixed When compiling Hex/Asm with PIC target and then calling FlashMcu, the window still closes
      immediately without a word...

== Release 5.3.0.5

    * Fixed flashMcu.bat and LIBRARIES_FOR ARM,AVR, PIC by Jose GILLES

    * Added: Microchip PIC12F675 8-pin packages

    * Added: Microchip PIC16F72 28-Pin PDIP, SOIC, SSOP

    * Added: Atmel AVR ATtiny10 6-Pin packages

== Release 5.3.0.4

    * Added: Menu Settings->Set Pull-up input resistors.
    All input pins try to set Pull-up registers by default.
    Attention: Not all ports have a pull-up resistor.
    See datasheets of the controller for details.

    * Added: PIC12F683 8-pin.

    * Added: ESP8266. ANSIC compiler only.

    * Fixed: Pins PB6 and PB7 are available for general I/O use when
    the internal clock source is selected.
    MCU list: ATmega8, ATmega48, ATmega88, ATmega168, ATmega328.
    Note: LDmicro not provided any clock source manipulation.

ldmicro/bits.h
#ifndef __BITS_H
#define __BITS_H

// clang-format off

#define BIT0   0
#define BIT1   1
#define BIT2   2
#define BIT3   3
#define BIT4   4
#define BIT5   5
#define BIT6   6
#define BIT7   7
#define BIT8   8
#define BIT9   9
#define BIT10 10
#define BIT11 11
#define BIT12 12
#define BIT13 13
#define BIT14 14
#define BIT15 15
#define BIT16 16
#define BIT17 17
#define BIT18 18
#define BIT19 19
#define BIT20 20
#define BIT21 21
#define BIT22 22
#define BIT23 23
#define BIT24 24
#define BIT25 25
#define BIT26 26
#define BIT27 27
#define BIT28 28
#define BIT29 29
#define BIT30 30
#define BIT31 31
#define BIT32 32

// clang-format on

//bitNumber to bitMask
#define _BV(bit) (1 << (bit))
#define bit(b) (1UL << (b))

#define bitRead(value, bit) (((value) >> (bit)) & 0x01) // Returns the value of the bit (0 or 1).
#define bitReadMask(value, bit) ((value)&bit(bit))      // Returns the mask value of the bit (0 or 0xXXXX).
#define bitSet(value, bit) ((value) |= (1UL << (bit)))
#define bitClear(value, bit) ((value) &= ~(1UL << (bit)))
#define bitWrite(value, bit, bitvalue) (bitvalue ? bitSet(value, bit) : bitClear(value, bit))

#define lowByte(w) ((uint8_t)((w)&0xff))
#define byte0(w) (lowByte(w))
#define byte1(w) ((uint8_t)((w) >> 8))
#define byte2(w) ((uint8_t)((w) >> 16))
#define byte3(w) ((uint8_t)((w) >> 24))

#endif
